File: day25/DayTwentyFive.java
import java.util.Queue;
import java.util.LinkedList;

public class DayTwentyFive
{
	class Node
	{
		int data;
		Node left;
		Node right;
		private Node(int data)
		{
			this.data = data;
			this.left = null;
			this.right = null;
		}
	}

	private int itr = 0;
	private int arr[] = {7,5,10,3,6,9,12};

	private Node btree = null;
	
	private Node createBtree()
	{
		for(int itr=0;itr<arr.length;itr++)
		{
			Node newnode = new Node(arr[itr]);
			btree = insertNode(btree,newnode);
		}
		return btree;
	}

	private Node insertNode(Node btree,Node newnode)
	{
		if(btree == null)
		{
			return newnode;
		}
		else if(newnode.data < btree.data)
		{
			btree.left = insertNode(btree.left,newnode);
		}
		else if(newnode.data > btree.data)
		{
			btree.right = insertNode(btree.right,newnode);
		}
		else if(newnode.data == btree.data)
		{
			System.out.print("Data Already Present");
			return btree;
		}
		return btree;
	}

	private int maxWidth(Node newnode)
	{
		Queue<Node> q = new LinkedList<Node>();
		q.add(newnode);
		int max_width = 0;
		while(q.isEmpty() == false)
		{
			if(max_width < q.size())
			{
				max_width = q.size();
			}

			Node temp = q.peek();
			q.remove();

			if(temp.left != null)
			{
				q.add(temp.left);
			}

			if(temp.right != null)
			{
				q.add(temp.right);
			}
		}
		return max_width;
	}

	public static void main(String args[])
	{
		DayTwentyFive tree = new DayTwentyFive();
		Node btree = tree.createBtree();

		System.out.print("maxWidth of given BianryTree => ");
		System.out.println(tree.maxWidth(btree));
	}
}

File: day25/day_twenty_five.cpp
# include <iostream>
# include <queue>

using namespace std;

typedef struct Node
{
	int data;
	struct Node* left;
	struct Node* right;
}node;

int itr = 0;
int arr[] = {7,5,10,3,6,9,12};
int size = sizeof(arr)/sizeof(arr[0]);


node* getNode()
{
	node* newnode = (node*)malloc(sizeof(node));
	return newnode;
}

void readNode(node* newnode)
{
	newnode->data = arr[itr];
	itr = itr + 1;
	newnode->left = NULL;
	newnode->right = NULL;
}

node* insertNode(node* btree,node* newnode)
{
	if(btree == NULL)
	{
		return newnode;
	}
	else if(newnode->data < btree->data)
	{
		btree->left = insertNode(btree->left,newnode);
	}
	else if(newnode->data > btree->data)
	{
		btree->right = insertNode(btree->right,newnode);
	}
	else if(btree->data == newnode->data)
	{
		cout << "data already exist" << endl;
		return btree;
	}
	return btree;
}

node* createBtree()
{
	node* btree = NULL;
	for(int i=0;i<size;i++)
	{
		node* newnode = getNode();
		readNode(newnode);
		btree = insertNode(btree,newnode);
	}
	return btree;
}

int maxWidth(node* newnode)
{
	queue<node*> q;
	q.push(newnode);
	int max_size = 0;
	while(q.empty() == false)
	{
		if(max_size < q.size())
		{
			max_size = q.size();
		}

		node* temp = q.front();
		q.pop();

		if(temp->left != NULL)
		{
			q.push(temp->left);
		}

		if(temp->right != NULL)
		{
			q.push(temp->right);
		}
	}
	
	return max_size;
}

int main()
{
	node* btree = createBtree();

	cout << "maxWidth of BianryTree => ";
	cout << maxWidth(btree) << endl;

	return 0;
}